### PR TITLE
feat: flag to pull request create that allows update of existing PR

### PR DESCRIPTION
**Pull request by waciumawanjohi**
Automated use of jx-scm may lead to repeated calls to create a PR between `feature` branch and `main` branch. If an existing PR exists between these two branches, the create call will fail. Users may prefer that the existing PR is simply updated. These changes enable that behavior by adding an option `--allowUpdate` flag.

The new tests rely on code [in this PR](https://github.com/jenkins-x/go-scm/pull/324) in go-scm.

### CODE REVIEW

**Comment by ankitm123** [comment]
@waciumawanjohi rebase ur PR with the latest master/main, afte #22 is merged.

File: pkg/cmd/pull_request/create/create_test.go [hunk 1]
```diff
@@ -20,25 +20,69 @@ func TestCreatePullRequest(t *testing.T) {
...
-	prs, _, err := scmClient.PullRequests.List(context.TODO(), fullName, scm.PullRequestListOptions{})
-	assert.NoError(t, err, "failed to list pull requests")
-	assert.Equal(t, 1, len(prs))
+	prs := assertExpectedPRFound(t, o, fullName, scmClient, err)
+
 	assert.Equal(t, prs[0].Title, o.Title, "title not properly set")
 	assert.Equal(t, prs[0].Body, o.Body, "body not properly set")
+}
+
+func createPRAgainWithoutAllowUpdate(t *testing.T, o *create_pr.Options, fullName string, scmClient *scm.Client, err error) {
```
Comment (ankitm123) [inline pkg/cmd/pull_request/create/create_test.go:49]: may be use table driven tests instead?

File: pkg/cmd/pull_request/create/create.go [hunk 2]
```diff
@@ -120,3 +147,24 @@ func (o *Options) Run() error {
 
 	return nil
 }
+
+func updateNecessary(head string, base string, updateAllowed bool, scmClient *scm.Client, ctx context.Context, fullName string) (bool, int) {
+	if !updateAllowed {
+		return false, 0
+	}
+	pullRequestListOptions := scm.PullRequestListOptions{Page: 1, Size: 5000, Open: true, Closed: false}
```
Comment (ankitm123) [inline pkg/cmd/pull_request/create/create.go:155]: Why 5000? Bit bucket for example sets the max at 100, so anything greater than that will result in bad request.
See: https://github.com/jenkins-x/go-scm/pull/319/files#diff-d11bfd98e6b72330250844f3d1b2a7b94e88a4245870e9eb8bfcb58e2bef3918R48

**Review by ankitm123 — COMMENTED** [review]
Few comments ...

**Comment by ankitm123** [comment]
/hold
some minor comments ...

File: pkg/cmd/pull_request/create/create.go (hunk 2 above)
Comment (waciumawanjohi) [reply pkg/cmd/pull_request/create/create.go:155]: > Why 5000?
I wanted a value that could reasonably assure that it wouldn't miss open PRs, particularly as the use case for this feature is an automated system (which could create many pull requests).

> Bit bucket for example sets the max at 100
That is great context on constraints. Will noodle.

Comment (ankitm123) [reply pkg/cmd/pull_request/create/create.go:155]: Curious, should it be a configurable parameter from the cli? That way, users of different scm providers can set whichever value they want (according to the scm limits)

**Comment by waciumawanjohi** [comment]
> Remove the hold after you have figured out if 5000 will break anything for other providers (including github/gitlab).

I've been able to create a pull request and then update that pull request for repositories in both github and gitlab using the `--allow-update` flag. Is that sufficient? Or is there a list of SCMs that need manual testing?

**Comment by ankitm123** [comment]
> Or is there a list of SCMs that need manual testing

github and gitlab should be fine (we anyways test with only github atm with automated tests). It would be good to test with bitbucket cloud (did u check the limits), but we can look at that later.

File: pkg/cmd/pull_request/create/create.go (hunk 2 above)
Comment (waciumawanjohi) [reply pkg/cmd/pull_request/create/create.go:155]: > Curious, should it be a configurable parameter from the cli? That way, users of different scm providers can set whichever value they want (according to the scm limits)

I don't think any user would understand why they would pass a page and size flag to a command to create a pull request.

Indeed, this make me wonder if the go-scm API is correct here. As a consumer of go-scm I expect to receive a slice of pull request objects. I would understand passing a length for that list (e.g. "Please provide me the most recent 50 or 500 PRs"). But questions of page or size limits should be the concern of the code within go-scm. Should the [PullRequestListOptions](https://github.com/jenkins-x/go-scm/blob/3b877743ecdbe91790294fdf6592e2b3c85380a2/scm/pr.go#L73) be changed?

Alternatively, should the provider specific code for listing PRs in go-scm handle `page` and `size` values of 0, setting them to reasonable defaults for the given provider?

**Comment by waciumawanjohi** [comment]
> It would be good to test with bitbucket cloud (did u check the limits), but we can look at that later.

I've been trying to figure out how to work with bitbucket cloud at all. What is your workflow for generating a personal access token for bitbucket cloud?
- `Repository settings -> Access keys` is about adding SSH keys
- There are instructions for OAuth, but that doesn't seem applicable
- I created an `app password` [here](https://bitbucket.org/account/settings/app-passwords/), but using it returns `error: failed to validate options: failed to validate options: failed to find current user: Not Authorized`

**Comment by ankitm123** [comment]
The version of go-scm in this project is ancient. There were a few fixes that went in, that fix a few issues with bitbucket cloud. I also use app password when playing around with bitbucket cloud. Normally the interaction happens using lighthouse with go-scm. I have not used jx-scm before (did not even know this repo existed), but will give it a try with bitbucket cloud tonight, and update this thread. Can you send me the command you run to verify it with other cloud providers?

**Comment by waciumawanjohi** [comment]
I decided to stop fighting the pagination and just use it. So the ridiculously large `size` value is gone.

One open question is how the different scm providers use that `close` value. E.g. Let's say that you have a repo where:
- PR #1 (least recent) is open.
- PR #2-#11 are closed.
- PR #12 is open.

What will be the behavior when calling
```
scm.PullRequestListOptions{Page: page, Size: 10, Open: true, Closed: false}
```
?

Will it return the 2 open PRs in the single page? Or will it return PR #12 and not PR #1 (because PR #1 is outside the pagination).

I tested manually against Github and it returns the 2 open PRs. But I don't know about the implementation of all other SCMs.

**Comment by waciumawanjohi** [comment]
My test command is:

```
./build/jx-scm pull-request create \
  --kind github \
  --server https://github.com/ \
  --token SOMETOKEN123 \
  --owner waciumawanjohi \
  --name demo-hello-world \
  --head hey \
  --title "It works, huzzah" \
  --body "Look here" \
  --base main \
  --allow-update
```

Run twice, it should succeed both times.

**Comment by waciumawanjohi** [comment]
This PR is blocked on this fix in go-scm:
https://github.com/jenkins-x/go-scm/pull/328

**Pull request merged**